### PR TITLE
Implement FlashListenerAPI for Wallet integration

### PR DESCRIPTION
**Pull request by AndrejMitrovic**
This will allow integrating the Flash node with a Wallet or other front-end client.

### CODE REVIEW

File: source/agora/flash/Node.d [hunk 1]
```diff
@@ -374,6 +405,12 @@ public abstract class FlashNode : ControlFlashAPI
 
         foreach (_, chan; this.channels)
             this.network.addChannel(chan.conf);
+
+        if (this.conf.listener_address.length != 0)
+            this.listener = this.getFlashListenerClient(
+                this.conf.listener_address, Duration.init);
+        else  // avoid null checks & segfaults
+            this.listener = new BlackHole!FlashListenerAPI();
```
Comment (AndrejMitrovic) [inline source/agora/flash/Node.d:413]: Notes about this:

- Either we use the code in the PR
- Or we make `listener` optional and we always do `&& this.listener !is null` everywhere.
- Or we make `listener` a required option so it must not be null.

File: source/agora/flash/Node.d [hunk 2]
```diff
@@ -834,12 +893,29 @@ public abstract class FlashNode : ControlFlashAPI
...
         }
 
         foreach (payment_hash; rev_htlcs)
+        {
             if (auto invoice = payment_hash in this.invoices)
-                this.payInvoice(*invoice); // Retry
+            {
+                ErrorCode error;
+                if (auto errors = payment_hash in this.payment_errors)
+                    error = (*errors)[0].err;  // pick first known reason
```
Comment (omerfirmak) [inline source/agora/flash/Node.d]: This should  be the last known reason as we queue new failure reasons to the end.
But still, we may report wrong failure reason to the wallet. As error are propagated separately, there is a race between the error message and HTLC being revoked.

Comment (AndrejMitrovic) [reply source/agora/flash/Node.d]: Yeah I had the same thoughts, not sure what is the best approach. I think I will change it to use the last error then.

**Comment by AndrejMitrovic** [comment]
@omerfirmak let me know if you have any additional comments and I'll fix them up in another PR.

**Pull request merged**